[docs] In-App Purchases (#4607)
expo/expo: docs/pages/versions/unversioned/sdk/in-app-purchases.md

---
title: InAppPurchases
---

An API to accept payments for in-app products. Internally relies on the [Google Play Billing](https://developer.android.com/google/play/billing/billing_library_overview) library on Android and [Storekit](https://developer.apple.com/documentation/storekit?language=objc) framework on iOS.

## Installation

This module is currently only available in the [bare](../../introduction/managed-vs-bare/#bare-workflow) workflow.

Note that in-app purchases require physical devices to work and therefore cannot be tested on simulators.

## Setup

### iOS

In order to use the In-App Purchases API on iOS, you’ll need to sign the [Paid Applications Agreement](https://help.apple.com/app-store-connect/#/devb6df5ee51) and set up your banking and tax information. You also need to enable the [In-App Purchases capability](https://help.apple.com/xcode/mac/current/#/dev88ff319e7) for your app in Xcode.

> Next, create an entry for your app in [App Store Connect](https://appstoreconnect.apple.com/) and configure your in-app purchases, including details such as name, pricing, and description that highlight the features and functionality of your in-app products. Make sure each product's status says `Ready to Submit`, otherwise it will not be queryable from within your app while you are testing. Be sure to add any necessary metadata to do so. You do not need to actually submit your app or its products to test purchases in sandbox mode.

Author: Apple requires you to upload a screenshot of the product and add review notes for it. It wasn't really clear that that was necessary when I was starting out which is why I mentioned it. I added a not to clarify what I meant.

On iOS, the refresh boolean is ignored and it returns the purchase history in the same way that `connectAsync` does. An important thing to note is that on iOS, Storekit actually creates a new transaction object every time you restore completed transactions, therefore the `purchaseTime` and `orderId` may be inaccurate if it's a restored purchase. If you need the original transaction's information you can use `originalPurchaseTime` and `originalOrderId`, but those will be 0 and an empty string respectively if it is the original transaction.

#### Arguments

- **refresh (_boolean_)** -- A boolean that indicates whether or not you want to make a network request to sync expired/consumed purchases and those on other devices (Android only)

#### Returns

A `Promise` that is resolved with a `QueryResponse` that contains an array of `Purchase` objects.

#### Example
```javascript
const { responseCode, results } = await getPurchaseHistoryAsync();
if (responseCode === ResponseCode.OK) {
    results.forEach(result => {
        // Handle purchase history
    })
}
```

### `InAppPurchases.purchaseItemAsync(productId: string, oldItem?: string)`

> Initiates the purchase flow to buy the item associated with this productId. This function is void and the result must be handled in the callback that you passed in to `setPurchaseListener`. This will display a prompt to the user which will allow them to either buy the item or cancel the purchase.

Author: Yep, will do.

#### Arguments

- **productId (_string_)** -- The product ID of the item you want to buy.

- **oldItem (_string_)** -- The product ID of the item you want to replace with this new purchase, typically a subscription such as when a user upgrades from monthly to yearly (Android only).

#### Example
```javascript
renderItem(item) {
    // Render product details with a "Buy" button
    return (
        <View key={item.productId}>
            ...
            <View style={styles.buttonContainer}>
                <Button title="Buy" onPress={() => purchaseItemAsync(item.productId)} />
            </View>
            ...
        </View>
    );
}
```

### `InAppPurchases.finishTransactionAsync(purchase: Purchase, consumeItem: boolean)`

Marks a transaction as completed. This must be called on successful purchases only after you have successfully processed the transaction and unlocked the functionality purchased by the user.

> On Android, this will either "acknowledge" or "consume" the purchase depending on the value of `consumeItem`. Consuming a purchase allows it to be bought more than once. You cannot buy an item again until it's consumed. Both consuming and acknowledging let Google know that you are done processing the transaction. Acknowledging indicates that this is a one time purchase (e.g. premium upgrade), rather than a consumable. If you do not acknowledge or consume a purchase within three days, the user automatically receives a refund, and Google Play revokes the purchase.

Reviewer: ```suggestion
On Android, this will either "acknowledge" or "consume" the purchase depending on the value of `consumeItem`. Acknowledging indicates that this is a one time purchase (e.g. premium upgrade), whereas consuming a purchase allows it to be bought more than once. You cannot buy an item again until it's consumed. Both consuming and acknowledging let Google know that you are done processing the transaction. If you do not acknowledge or consume a purchase within three days, the user automatically receives a refund, and Google Play revokes the purchase.
```

| --------------------- | --------- | --------- | ----------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| responseCode                 | _ResponseCode_  | both      | The response code from a query or purchase.                                  | `ResponseCode.OK`, `ResponseCode.USER_CANCELED`, `ResponseCode.ERROR`, `ResponseCode.DEFERRED` (iOS only)                                                                                                                                                                                                                                                                                                                                                                   |

> | results   | _Purchase[] or ItemDetails[]_ | Android      | The array containing the `Purchase` or `ItemDetails` objects requested depending on the method.           |                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer: I thought results was both platforms, but for iOS it would just contain one item?

Author: No, both return an array for consistency but iOS will typically only have one entry in it, at least when you're making a purchase.

Reviewer: Right, but it lists `Android` under platforms instead of `both`- since both return an array (even though iOS is an array of 1), shouldn't it say `both` under Platforms?

Author: Oh whoops, that was totally a mistake. Good catch 👍

| type                  | _string_  | both       | The type of the purchase. Note that this is not very accurate on iOS as this data is only available on iOS 11.2 and higher and non-renewable subscriptions always return `inapp`                                      |  `inapp`, `subs`                                                                                                                                                                                          |
| subscriptionPeriod             | _string_ | both   | The length of a subscription period, specified in ISO 8601 format. On iOS, `inapp` purchases and non-renewable subscriptions return `P0D`. On Android, `inapp` purchases return an empty string.      | `P0D`, `P6W`, `P3M`, `P6M`, `P1Y` |

## Enum Types

### `InAppPurchases.ResponseCode`

- **`ResponseCode.OK`** - Response returned successfully.
- **`ResponseCode.USER_CANCELED`** - User canceled the purchase.
- **`ResponseCode.ERROR`** - An error occurred. Check the `errorCode` for additional details.
- **`ResponseCode.DEFERRED`** - Purchase was deferred. (iOS only)

> ### `InAppPurchases.ErrorCode`

Reviewer: ❤️ this section

On iOS, the refresh boolean is ignored and it returns the purchase history in the same way that `connectAsync` does. An important thing to note is that on iOS, Storekit actually creates a new transaction object every time you restore completed transactions, therefore the `purchaseTime` and `orderId` may be inaccurate if it's a restored purchase. If you need the original transaction's information you can use `originalPurchaseTime` and `originalOrderId`, but those will be 0 and an empty string respectively if it is the original transaction.

#### Arguments

- **refresh (_boolean_)** -- A boolean that indicates whether or not you want to make a network request to sync expired/consumed purchases and those on other devices (Android only)

#### Returns

A `Promise` that is resolved with a `QueryResponse` that contains an array of `Purchase` objects.

#### Example
```javascript
const { responseCode, results } = await getPurchaseHistoryAsync();
if (responseCode === ResponseCode.OK) {
    results.forEach(result => {
        // Handle purchase history
    })
}
```

> ### `InAppPurchases.purchaseItemAsync(productId: string, oldItem?: string)`

Reviewer 2: if the function returns void, should it still be called `Async`? if it actually returns a promise, we should document when the promise is expected to resolve. (e.g. resolves as soon as the network request to the store completes successfully)

Author: It's async because it resolves a promise natively. It doesn't return anything since the response is handled by the callback in `setPurchaseListener`. Not sure what the convention is there but I thought every function that makes a native method call should end in `Async`.

> Initiates the purchase flow to buy the item associated with this productId. This function is void and the result must be handled in the callback that you passed in to `setPurchaseListener`. This will display a prompt to the user which will allow them to either buy the item or cancel the purchase.

Reviewer 3: ```suggestion
Initiates the purchase flow to buy the item associated with this `productId`. This will display a prompt to the user which will allow them to either buy the item or cancel the purchase. When the purchase completes, the result must be handled in the callback that you passed in to `setPurchaseListener`. 
```

later:

```
#### Returns

 A `Promise` that is resolved when the purchase flow starts. To get the actual result of the purchase, you must handle purchase events with `setPurchaseListener`.
```

> Remember, you have to query an item's details via `getProductsAsync` and set the purchase listener before you attempt to buy an item.

Reviewer 2: what happens if you don't?

Author: Don't call `getProductsAsync` on an item before `purchaseItemAsync`? It rejects the promise and throws.

#### Arguments

- **productId (_string_)** -- The product ID of the item you want to buy.

> - **oldItem (_string_)** -- The product ID of the item you want to replace with this new purchase, typically a subscription such as when a user upgrades from monthly to yearly (Android only).

Reviewer 2: what is the effect of setting this parameter? does it overwrite the `oldItem` in the purchase history or something?

Author: It replaces a previous purchase. A good example is when a user buys a monthly subscription and then describes to upgrade to a yearly subscription for the same thing. Setting `oldItem` would prevent the user from being double charged.

### `InAppPurchases.purchaseItemAsync(productId: string, oldItem?: string)`

> Initiates the purchase flow to buy the item associated with this productId. This function returns null as the result must be handled in the callback that you passed in to `setPurchaseListener`. This will display a prompt to the user which will allow them to either buy the item or cancel the purchase.

Reviewer 3: The TS source code returns `Promise<void>`, which is conceptually correct here. Since the body of the TS function just awaits the native method (rather than `return await`) -- the returned promise definitely resolves to undefined (`Promise<void>`)

Also, the promise's value doesn't really matter -- just the fact that this method is async and returns a promise to await is more important. So phrasing the docs to be focused more on awaiting the promise and the timing of when it's resolved helps highlight what's important here, ex:

"This function returns a void promise that resolves when <describe condition here>. 

---

Separately, I find this part confusing: 
> ...that you passed in to `setPurchaseListener`. This will display a prompt...

Does "This" refer to `purchaseItemAsync` or the purchase listener?

Author: Changed in latest commits to:
> Initiates the purchase flow to buy the item associated with this `productId`. This will display a prompt to the user that will allow them to either buy the item or cancel the purchase. When the purchase completes, the result must be handled in the callback that you passed in to `setPurchaseListener`.

The prompt is displayed as a result of calling `purchaseItemAsync`. The purchase listener doesn't receive the result until the user clicks "Buy" or "Cancel".

[Apple](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Chapters/Subscriptions.html) and [Google](https://developer.android.com/google/play/billing/billing_subscriptions) both have their own workflows for dealing with subscriptions. In general, you can deal with them in the same way you do one-time purchases but there are caveats including if a user decides to cancel before the expiration date. To check the status of a subscription, you can use the [Google Play Developer](https://developers.google.com/android-publisher/api-ref/purchases/subscriptions/get) API on Android and the [Status Update Notifications](https://developer.apple.com/documentation/storekit/in-app_purchase/enabling_status_update_notifications) service on iOS, both of which are REST APIs.

#### Arguments

- **productId (_string_)** -- The product ID of the item you want to buy.

- **oldItem (_string_)** -- The product ID of the item that the user is upgrading or downgrading from. This is mandatory for replacing an old subscription such as when a user upgrades from a monthly subscription to a yearly one for the identical content (Android only).

#### Returns

> A `Promise` that resolves with `null` when the purchase is done processing.

Reviewer 3: null -> undefined

Author: You might be reviewing an outdated commit. I changed it to 

> A `Promise` that resolves when the purchase is done processing. To get the actual result of the purchase, you must handle purchase events inside the `setPurchaseListener` callback.

### `InAppPurchases.disconnectAsync()`

Disconnects from the app store and cleans up memory internally. Call this when you are done using the In-App Purchases API in your app.

No other methods can be used until the next time you call `connectAsync`.

#### Returns

> A promise that resolves with `null` when finished disconnecting.

Reviewer 3: null -> undefined

Author: Same as above. Changed to:
> A `Promise` that resolves when finished disconnecting.
